dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java

```diff
@@ -0,0 +1,167 @@
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.Event;
+import com.dat3m.dartagnan.program.event.EventVisitor;
+import com.dat3m.dartagnan.program.event.RegReader;
+import com.dat3m.dartagnan.program.event.RegWriter;
+import com.dat3m.dartagnan.program.event.core.Alloc;
+import com.dat3m.dartagnan.program.event.core.Local;
+import com.dat3m.dartagnan.program.memory.MemoryObject;
+
+import java.math.BigInteger;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.dat3m.dartagnan.expression.integers.IntBinaryOp.*;
+import static com.dat3m.dartagnan.expression.integers.IntCmpOp.*;
+
+/*
+    This pass tries to remove unnecessary NULL checks in a very simple manner:
+    it tries to figure out the signedness of expressions and if they are always positive, then they cannot be NULL.
+
+    The signedness is determined as follows:
+      - Registers written to by Allocs as well as MemoryObject expressions are always positive (allocations always succeed!)
+      - Positive constants are positive (duh).
+      - Some operations (like addition) preserve positiveness.
+
+    TODO: The pass is very naive: it globally assigns a signedness to registers rather than per program point.
+ */
+public class RemoveDeadNullChecks implements FunctionProcessor {
+
+    private RemoveDeadNullChecks() { }
+
+    public static RemoveDeadNullChecks newInstance() { return new RemoveDeadNullChecks(); }
+
+    private enum Sign {
+        UNKNOWN,
+        NON_NEG,
+        POS;
+
+        static Sign meet(Sign a, Sign b) {
+            return a.compareTo(b) >= 0 ? b : a;
+        }
+    }
+
+    @Override
+    public void run(Function function) {
+
+        // Collects signs of registers.
+        final SignChecker signChecker = new SignChecker();
+        function.getEvents().forEach(e -> e.accept(signChecker));
+
+        // Simplify null checks on always-positive (and hence non-null) expressions.
+        final NullCheckReplacer replacer = new NullCheckReplacer(signChecker);
+        function.getEvents(RegReader.class).forEach(reader -> reader.transformExpressions(replacer));
+
+    }
+
+    private static class NullCheckReplacer extends ExprTransformer {
+
+        private final SignChecker signChecker;
+
+        private NullCheckReplacer(SignChecker signChecker) {
+            this.signChecker = signChecker;
+        }
+
+        @Override
+        public Expression visitIntCmpExpression(IntCmpExpr cmp) {
+            if (cmp.getRight() instanceof IntLiteral lit && lit.isZero() && cmp.getLeft().accept(signChecker) == Sign.POS) {
+                // Simplify "expr cop 0" if <expr> is known to be positive.
+                final IntCmpOp op = cmp.getKind();
+                final boolean valueOnNonNull = op == NEQ || op == GT || op == GTE || op == UGT || op == UGTE;
+                return ExpressionFactory.getInstance().makeValue(valueOnNonNull);
+            }
+            return cmp;
+        }
+    }
```

Comment:
Would it make sense to have a dual pattern for computing expression that are definitely negative?

Reply:
Possible but probably useless.